internal/config/variables.go

```diff
@@ -0,0 +1,420 @@
+				// 	sev = hcl.DiagError
+				// }
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: sev,
+					Summary:  "Undefined variable",
+					Detail: fmt.Sprintf("A %q variable was set but was "+
+						"not found in known variables. To declare "+
+						"variable %q, place this block in your "+
+						"waypoint.hcl file.",
+						name, name),
+					Context: attr.Range.Ptr(),
+				})
+				continue
+			}
+
+			val, moreDiags := attr.Expr.Value(nil)
+			diags = append(diags, moreDiags...)
+
+			if variable.Type != cty.NilType {
+				var err error
+				val, err = convert.Convert(val, variable.Type)
+				if err != nil {
+					diags = append(diags, &hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Invalid value for variable",
+						Detail:   fmt.Sprintf("The value for %s is not compatible with the variable's type constraint: %s.", name, err),
+						Subject:  attr.Expr.Range().Ptr(),
+					})
+					val = cty.DynamicVal
+				}
+			}
+
+			variable.Values = append(variable.Values, VariableAssignment{
+				Source: "repofile",
+				Value:  val,
+				Expr:   attr.Expr,
+			})
+		}
+	}
+
+	// Finally we process values given explicitly on the command line.
+	for _, sv := range serverVars {
+		variable, found := (*variables)[sv.Name]
+		if !found {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Undefined server variable",
+				Detail: fmt.Sprintf("A %q variable was set in the UI "+
+					"and stored on the server, but was not found in "+
+					"known variables. To declare variable %q, place "+
+					"this block in your waypoint.hcl file.",
+					sv.Name, sv.Name),
+			})
+			continue
+		}
+
+		var expr hclsyntax.Expression
+		switch sv.Value.(type) {
+
+		case *pb.Variable_Hcl:
+			value := sv.Value.(*pb.Variable_Hcl).Hcl
+			fakeFilename := fmt.Sprintf("<value for var.%s from server>", sv.Name)
+			expr, diags = hclsyntax.ParseExpression([]byte(value), fakeFilename, hcl.Pos{Line: 1, Column: 1})
+
+		case *pb.Variable_Str:
+			value := sv.Value.(*pb.Variable_Str).Str
+			expr = &hclsyntax.LiteralValueExpr{Val: cty.StringVal(value)}
+		}
+
+		val, valDiags := expr.Value(nil)
+		diags = append(diags, valDiags...)
+		if valDiags.HasErrors() {
+			return diags
+		}
+
+		if variable.Type != cty.NilType {
+			var err error
+			val, err = convert.Convert(val, variable.Type)
+			if err != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid argument value for server-stored variable",
+					Detail:   fmt.Sprintf("The received arg value for %s is not compatible with the variable's type constraint: %s.", sv.Name, err),
+					Subject:  expr.Range().Ptr(),
+				})
+				val = cty.DynamicVal
+			}
+		}
+
+		variable.Values = append(variable.Values, VariableAssignment{
+			Source: "server",
+			Value:  val,
+			Expr:   expr,
+		})
+	}
+
+	return diags
+}
+
+func (variables *Variables) SortPrecedence() ([]*pb.Variable, error) {
```

Comment:
Also, this seems super important and should be tested. This is like, _the_ logic behind what values are used, I think.

Reply:
For sure. This was the point where I stopped and decided to push up what I had. This function is only half done; I think I noted that where I call it, rather than as a comment here, so that's on me. I'm not totally sure how I want this function to work yet.

Reply:
Ok as mentioned in an earlier comment, I've decided to take the sort of precedence out of an explicit function call, and evaluate values in precedence order. Again, not tied to either implementation (but FYI this function doesn't exist anymore)